Use `Timer.periodic` in the `Resampler`

jack-herrmann opened the pull request:
Fixes #264

jack-herrmann committed:
Resampler emits sample as regularly as possible using Timer.periodic

Signed-off-by: Jack <jack.herrmann@gmx.de>

shsms commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -461,15 +463,33 @@ async def resample(self, *, one_shot: bool = False) -> None:
+            tolerance = timedelta(
+                seconds=self._config.resampling_period.total_seconds() / 10.0
+            )
+            if timer_error > tolerance:
```
I think all this code can go away now and we can just use the drift that the timer returns.

```suggestion
        # We use a tolerance of 10% of the resampling period
        tolerance = timedelta(
            seconds=self._config.resampling_period.total_seconds() / 10.0
        )
        async for drift in Timer.periodic(
            timedelta(seconds=self._config.resampling_period.total_seconds())
        ):
            if drift > tolerance:

jack-herrmann replied:
now and time_error also need to be updated in the logger message, unfortunately I saw this after accepting the change

llucax commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -461,15 +463,33 @@ async def resample(self, *, one_shot: bool = False) -> None:
+        tolerance = timedelta(
+            seconds=self._config.resampling_period.total_seconds() / 10.0
+        )
+        async for drift in Timer.periodic(
+            timedelta(seconds=self._config.resampling_period.total_seconds())
+        ):
```
This timer needs to be a instance member, otherwise the `drift` will always be zero, as you creating the timer on each resample call. For the Resampling actor this is not an issue because `resample()` is called once and then we loop on it, but in the tests we use `one_shot` and in that case we call `resample()` multiple times, thus triggering the permanent zero drift.

Also the tests need to be adjusted to also do a `asyncio.sleep()` with the same time as `fake_time.shift()` to make sure the loop monotonic clock (which is the one the new channels timer uses) is also advanced to produce a drift, otherwise the timer will always trigger at the interval time.

llucax reviewed (changes requested):
I will send a PR to your branch with a fix, as it is not trivial to explain in comments.

llucax committed:
Preserve the state of the resampling timer

The resampling timer needs to preserve its state when using in single
shot mode, so it needs to be an instance member.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Fix resample timer drifting tests

Now that we use the new channels timer, we need to adjust the tests to
account for the loop time too. We also need to account for the lazy
timer start.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
There is another issue with this change, the `align_to` is also broken, and fixing it is not all that trivial. Adding alignment to the channels timer is hard because it doesn't know anything about `datetime` and we want to align to a `datetime`.

So to align the timer we need to actually have an initial `Timer.timeout()` to just wait until the next aligned window and then start the `Timer.periodic()` after that. Having channels (having https://github.com/frequenz-floss/frequenz-channels-python/issues/116 would make this **much** easier), but for now I think we need to refactor things a bit to make it work. I'll still send a proposal fix but it is taking me a bit more time than expected.

llucax commented:
I'm starting to think that a periodic timer is not a good fit for this :grimacing: 

If I add a timer to align, but **that** timer drifts too much, then what? :grimacing: :gun: 

We are really getting into a mess here, this issue was created before we realized we needed to align the resampling window, and the alignment only makes sense using a `datetime`.

We either need to live with the fact that the periodic time will be a bit shifted (we can still emit the timestamps for samples and do all the calculations with the desired precise window we want, it is only that the sample will be sent a bit later than that timestamp) and don't care about **alignment of the timer**, or we should probably leave things as they are.

And to be honest, I'm leaning more towards leave what we have, it is less effort and probably more accurate (less chances of having a considerable shift between the timer triggering and the timestamp being emitted).

llucax committed:
Properly align the timer to align_to

By using the Timer.periodic, the alignment was lost. We now make the
timer aligned to the align_to configuration by delaying its start.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Always wait for at least one resampling period

When starting the resampler, if the time is not aligned to `align_to`,
it might happen that the first resampling window is too small, so it
will produce a bad first sample.

This commit ensures that the first window includes at least all the
samples of a full window, so the first resampling will be performed at
least one period after it was started (one period plus the time needed
for alignment).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Proposed hacky fix for this: https://github.com/jh2007github/frequenz-sdk-python/pull/1

I would leave this open and implement the hack properly in `Timer` and only merge it when we upgraded the SDK to use the version with a `start_delay: timedelta` option, since there is no real urgency to merge it.

llucax committed:
Rename receiver/sender variables to make more sense

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax reviewed (dismissed):
We decided to merge this as is and update when the `Timer` gets the new `start_delay` option.

llucax commented:
BTW, this needs someone else's approval, as I made the latest commits in this PR :innocent:

llucax commented:
Setting auto-merge.

llucax commented:
* https://github.com/frequenz-floss/frequenz-sdk-python/issues/556

llucax commented:
@jh2007github not sure what's going on here with:

![image](https://github.com/frequenz-floss/frequenz-sdk-python/assets/1031485/4c176fa2-0f51-4f32-a2c4-b8adc79675e7)

llucax commented:
It would be good to merge this before releasing v0.23.0 as it is ready already, but the approvals keep getting dismissed by the base change.
